features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,1066 @@
+// offset        - [IN]   Offset in area.
+// len_bytes     - [IN]   Length in bytes.
+// buf           - [IN]   Data buffer.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_read_area(uint8_t area, uint32_t offset, uint32_t len_bytes, uint32_t *buf)
+{
+    return nvstore_int_flash_read(len_bytes, _flash_area_params[area].address + offset, buf);
+}
+
+// Write to flash, given area and offset.
+// Parameters :
+// area          - [IN]   Flash area.
+// offset        - [IN]   Offset in area.
+// len_bytes     - [IN]   Length in bytes.
+// buf           - [IN]   Data buffer.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_write_area(uint8_t area, uint32_t offset, uint32_t len_bytes, const uint32_t *buf)
+{
+    return nvstore_int_flash_write(len_bytes, _flash_area_params[area].address + offset, buf);
+}
+
+// Erase a flash area, given area.
+// Parameters :
+// area          - [IN]   Flash area.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_erase_area(uint8_t area)
+{
+    return nvstore_int_flash_erase(_flash_area_params[area].address, _flash_area_params[area].size);
+}
+
+// Calculate area address and size configuration (in case it's not determined by the user)
+// or validate it (in case it is)
+void NVStore::calc_validate_area_params()
+{
+    int num_sectors = 0;
+
+    size_t flash_addr = nvstore_int_flash_get_flash_start();
+    size_t flash_size = nvstore_int_flash_get_flash_size();
+    size_t sector_size;
+    int max_sectors = flash_size / nvstore_int_flash_get_sector_size(flash_addr) + 1;
+    size_t *sector_map = new size_t[max_sectors];
+
+    int area = 0;
+    size_t left_size = flash_size;
+#if AREA_PARAMS_USER_CONFIG
+    _flash_area_params[0].address = NVSTORE_AREA_1_ADDRESS;
+    _flash_area_params[0].size    = NVSTORE_AREA_1_SIZE;
+    _flash_area_params[1].address = NVSTORE_AREA_2_ADDRESS;
+    _flash_area_params[1].size    = NVSTORE_AREA_2_SIZE;
+
+    int in_area = 0;
+    size_t area_size = 0;
+#endif
+    while (left_size) {
+        sector_size = nvstore_int_flash_get_sector_size(flash_addr);
+        //printf("Sector %3d: Address 0x%08x, size %6ld\n", num_sectors, flash_addr, sector_size);
+        sector_map[num_sectors++] = flash_addr;
+#if AREA_PARAMS_USER_CONFIG
+        // User configuration - here we validate it
+        if (_flash_area_params[area].address == flash_addr) {
+            in_area = 1;
+        }
+        if (in_area) {
+            area_size += sector_size;
+            if (area_size == _flash_area_params[area].size) {
+                area++;
+                if (area == NVSTORE_NUM_AREAS) {
+                    break;
+                }
+                in_area = 0;
+                area_size = 0;
+            }
+        }
+#endif
+        flash_addr += sector_size;
+        left_size -= sector_size;
+    }
+    sector_map[num_sectors] = flash_addr;
+
+#if AREA_PARAMS_USER_CONFIG
```

Comment:
Why are all areas not asserted if wrong?

Reply:
There is such an assertion in line 266. This function checks whether both areas are on aligned addresses, have aligned sizes and don't overlap. Area count is incremented for each legal area, so if count doesn't reach 2, we get an assertion. I'll add some comments there to clarify.